File: SDK Examples/Examples/Java/src/Counter_BufferedUpDownCounter/BufferedUpDownCounter.java
package Counter_BufferedUpDownCounter;

import java.awt.Component;
import java.awt.EventQueue;
import java.awt.SystemColor;
import java.awt.Color;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;

import javax.swing.JFrame;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;
import javax.swing.BorderFactory;
import javax.swing.DefaultListModel;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JTextField;
import javax.swing.JList;
import javax.swing.ListCellRenderer;
import javax.swing.ListSelectionModel;
import javax.swing.border.Border;
import javax.swing.border.LineBorder;
import javax.swing.SwingConstants;
import javax.swing.JButton;

import Automation.BDaq.*;
import Common.Global;

import org.eclipse.wb.swing.FocusTraversalOnArray;

import java.awt.Font;

import javax.swing.JComboBox;
import javax.swing.JTextArea;

public class BufferedUpDownCounter extends JFrame implements ActionListener{
	// define the serialization number.
	private static final long serialVersionUID = 1L;
	
	private JPanel contentPane;
	private JTextField txtChannelStatus;
	private JList<String> lstCounterValue;
	private JComboBox cbxDisplayedChannel;
	private DefaultListModel<String> model = new DefaultListModel<String>();
	private JButton btnStart;
	private JButton btnStop;
	private JButton btnConfigure;
	private int[] dataBuf;
	private int[][] chs_dataBuf;
	private String[] ch_status;
	private int eventChannel;
	private int displayChannel;
	private boolean [] channelEnable;
	private int sectionLength;
	private int sectionCount;
	private int samplesCount;
	private boolean isRunning;
	
	private boolean haveData;
	private int listCount;
	
	private final int MAX_ROW_COUNT = 10;
	
	ConfigureParameter configure = new ConfigureParameter();
	private ConfigureDialog configureDialog;
	private BufferedUdCounterCtrl bfdUdCounterCtrl = new BufferedUdCounterCtrl();
	private int channelCountMax;
	private JLabel lblNewLabel_2;
	
	/**
	 * 
	 * Build Date:2011-9-29 
	 * Author:Administrator
	 * Function Description: Launch the application.
	 */
	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					BufferedUpDownCounter frame = new BufferedUpDownCounter();
					frame.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	/**
	 * 
	 * Build Date:2011-9-29 
	 * Author:Administrator
	 * Function Description: Create the frame.
	 */
	public BufferedUpDownCounter() {
		setResizable(false);
		setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		setBounds(100, 100, 450, 536);
		addWindowListener(new WindowCloseActionListener());
		
		contentPane = new JPanel();
		contentPane.setBackground(SystemColor.control);
		contentPane.setBorder(new EmptyBorder(5, 5, 5, 5));
		setContentPane(contentPane);
		contentPane.setLayout(null);
		
		txtChannelStatus = new JTextField();
		txtChannelStatus.setHorizontalAlignment(SwingConstants.RIGHT);
		txtChannelStatus.setEditable(false);
		txtChannelStatus.setBounds(100, 109, 120, 21);
		contentPane.add(txtChannelStatus);
		txtChannelStatus.setColumns(10);
		
		lstCounterValue = new JList<String>();
		lstCounterValue.setBackground(SystemColor.controlHighlight);
		lstCounterValue.setBorder(new LineBorder(new Color(0, 0, 0)));
		lstCounterValue.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
		lstCounterValue.setBounds(25, 150, 230, 287);
		lstCounterValue.setAlignmentX(Component.RIGHT_ALIGNMENT);
		lstCounterValue.setFixedCellHeight(28);
		contentPane.add(lstCounterValue);
		
		btnStart = new JButton("Start");
		btnStart.setBounds(296, 165, 90, 23);
		btnStart.addActionListener(new ButtonStartActionListener());
		contentPane.add(btnStart);
		
		btnStop = new JButton("Stop");
		btnStop.setBounds(296, 226, 90, 23);
		btnStop.addActionListener(new ButtonStopActionListener());
		btnStop.setEnabled(false);
		contentPane.add(btnStop);
		
		btnConfigure = new JButton("Configure");
		btnConfigure.setBounds(296, 336, 90, 23);
		btnConfigure.addActionListener(new ButtonConfigureActionListener());
		contentPane.add(btnConfigure);
		
		// Add background image for the main Frame
		BackgroundPanel panel = new BackgroundPanel("Background.png", "background image!");
		panel.setBounds(269, 0, 175, 64);
		panel.setBackground(SystemColor.control);
		panel.setLayout(null);
		contentPane.add(panel);
		
		JLabel lblNewLabel = new JLabel("Channel:");
		lblNewLabel.setFont(new Font("Arial", Font.PLAIN, 15));
		lblNewLabel.setBounds(25, 70, 76, 29);
		contentPane.add(lblNewLabel);
		
		cbxDisplayedChannel = new JComboBox();
		cbxDisplayedChannel.setBounds(100, 74, 115, 21);
		cbxDisplayedChannel.addItemListener(new cbxDisplayedChannelItemListener());
		contentPane.add(cbxDisplayedChannel);
		
		JLabel lblNewLabel_1 = new JLabel("Status:");
		lblNewLabel_1.setFont(new Font("Arial", Font.PLAIN, 15));
		lblNewLabel_1.setBounds(25, 112, 71, 15);
		contentPane.add(lblNewLabel_1);
		
		lblNewLabel_2 = new JLabel("<html><body>Note:The example will display the first ten values on Data Ready Event every time.</body><html>");
		lblNewLabel_2.setBounds(25, 447, 283, 45);
		contentPane.add(lblNewLabel_2);
		contentPane.setFocusTraversalPolicy(new FocusTraversalOnArray(new Component[]{
				btnStart, btnStop, btnConfigure}));
	
		bfdUdCounterCtrl.addDataReadyListener(new DataReadyEventListener());
		bfdUdCounterCtrl.addStoppedListener(new StoppedEventListener());
		
		configureDialog = new ConfigureDialog(this);
		configureDialog.setModal(true);
		configureDialog.setVisible(true);
	}
	
	/**
	 * 
	 * Build Date:2011-9-29 
	 * Author:Administrator
	 * Function Description: This function is used to initialize the main frame.
	 */
	public void Initialization() {	
		//String array channelValues contents of the current channels' data in each channel.
		String[] channelValues = new String[MAX_ROW_COUNT];
		//set the main frame's title.
		this.setTitle("Buffered Up Down Counter - Run(" + configure.deviceName.substring(0, configure.deviceName.length() - 1) + ")");
		
		model.removeAllElements();
		for (int i = 0; i < channelValues.length; i++) {
			channelValues[i] = "";
			//channelValues[i] = String.valueOf(i);
			model.addElement(channelValues[i]);
		}
		lstCounterValue.setModel(model);
		lstCounterValue.setBorder(new LineBorder(new Color(0, 0, 0)));
		lstCounterValue.setCellRenderer(new ListViewCellRenderer());
		lstCounterValue.setVisible(true);
		
		txtChannelStatus.setText("");
		
		isRunning = false;
		haveData = false;
		
		ConfigureDevice();		
	}
	
	/**
	 * 
	 *Build Date:2011-9-29
	 *Author:Administrator
	 *Function Description: This function is used to configure the event counter device.
	 */
	private void ConfigureDevice(){
		try{
			//configure device
			bfdUdCounterCtrl.setSelectedDevice(new DeviceInformation(configure.deviceName));
			channelCountMax = bfdUdCounterCtrl.getFeatures().getChannelCountMax();
			channelEnable = new boolean [channelCountMax];
			ch_status = new String[channelCountMax];
			
			// ConfigureMode: ProfileConfigure / ManualConfigure
			if(configure.configureMode == 1){			
				String profilePath = null;
				profilePath = configureDialog.GetProfilePath();
				ErrorCode errorCode = bfdUdCounterCtrl.LoadProfile(profilePath);
				CheckError(errorCode);
				channelEnable = bfdUdCounterCtrl.getChannelMap();
				sectionLength = bfdUdCounterCtrl.getRecord().getSectionLength();
				sectionCount = bfdUdCounterCtrl.getRecord().getSectionCount();
				
			}else{
				System.arraycopy(configure.channelEnabled, 0, channelEnable, 0, configure.channelEnabled.length);
				sectionLength = configure.sectionLength;
				sectionCount = configure.sectionCount;
				
				// Device General Settings
				bfdUdCounterCtrl.setChannelMap(configure.channelEnabled);
				bfdUdCounterCtrl.setSampleClkSrc(configure.sampleClkSrc);
				bfdUdCounterCtrl.getRecord().setSectionLength(configure.sectionLength);
				bfdUdCounterCtrl.getRecord().setSectionCount(configure.sectionCount);
				
				for(int i = 0; i < channelCountMax; i++){
					if(channelEnable[i]){
						bfdUdCounterCtrl.getChannels()[i].setCountingType(configure.cntType);
					}
				}
				
				// Trigger Settings
				Trigger trg = bfdUdCounterCtrl.getTrigger();
				if(trg != null){
					trg.setSource(configure.triggerSource);
					trg.setAction(configure.triggerAction);
					trg.setEdge(configure.triggerEdge);
					trg.setLevel(configure.triggerLevel);
				}
			}
			
			//configure dataBuf 
			samplesCount = sectionLength;
			
			if(sectionCount != 0){
				samplesCount *= sectionCount;
			}		
			if(samplesCount <= 1){
				samplesCount = 2;
			}
			dataBuf = new int[samplesCount];
			chs_dataBuf = new int[channelCountMax][samplesCount];
			
			cbxDisplayedChannel.removeAllItems();	
			cbxDisplayedChannel.setEnabled(false);
			//enum enabled channels for cbxEventChannel
			for(int i = 0; i < channelCountMax; i++){
				if(channelEnable[i]){
					cbxDisplayedChannel.addItem(String.valueOf(i));
				} 
			}
			cbxDisplayedChannel.setEnabled(true);
			cbxDisplayedChannel.setSelectedIndex(0);	
			
		}catch(Exception ex){
			ShowMessage("Sorry, there're some errors occured: " + ex.toString());
		}
		for(int i = 0; i < channelCountMax; i++){
			if(channelEnable[i]){
				ChannelStatusStringChange(i, "Ready");
			}
		}
		 displayChannel = Integer.parseInt(cbxDisplayedChannel.getSelectedItem().toString());
		 DisplayChannelStatus(displayChannel);
	}
	
	/**
	 * 
	 * Build Date:2011-9-29 
	 * Author:Administrator
	 * Function Description: This function is overload the interface ActionListener,and 
	 * 						 is implementation need this function!
	 */
	@Override
	public void actionPerformed(ActionEvent e) {

	}
	
	/**
	 * 
	 * Build Date:2011-9-29 
	 * Author:Administrator
	 * Function Description: If some errors occurred, Show the error code to the users.
	 * 
	 * @param message:the message shown to users!
	 */
	protected void ShowMessage(String message) {
		JOptionPane.showMessageDialog(this, message, "Warning MessageBox",
				JOptionPane.WARNING_MESSAGE);
	}
	
	/**
	 * 
	 * @author Administrator 
	 * Class Function Description: check errorcode
	 */
	protected void CheckError(ErrorCode errorCode)
	{
		if(Global.BioFaild(errorCode)){
			ShowMessage("Sorry, there're some errors occred, ErrorCode: " + errorCode.toString());
		}		
	}
	
	protected void ChannelStatusStringChange(int channel, String str)
	{
		ch_status[channel] = str;
	}
	
	protected void DisplayChannelStatus(int channel)
	{
		txtChannelStatus.setText(ch_status[channel]);
	}
	
	private void DisplayChannelValue(int channel)
	{	
		listCount = MAX_ROW_COUNT;
		if(sectionLength < MAX_ROW_COUNT){
			listCount = sectionLength;
		}
		
		if(isRunning | haveData){
			for(int i = 0; i < listCount; i++){
				model.set(i, String.valueOf(chs_dataBuf[channel][i]));
			}
		}
	}
	
	protected void CopyChannelData(int channel)
	{
		int ch = channel;
		System.arraycopy(dataBuf, 0, chs_dataBuf[ch], 0, dataBuf.length); 
	}
	
	/**
	 * 
	 * @author Administrator 
	 * Class Function Description: this class is use to draw each cell of the JList object
	 */
	class ListViewCellRenderer extends JLabel implements ListCellRenderer {
		/**
		 * define the serialization number.
		 */
		private static final long serialVersionUID = 1L;

		public Component getListCellRendererComponent(JList list, Object value, int index,
				boolean isSelected, boolean cellHasFocus) {

			if (value != null) {
				String text = value.toString();
				setText(text);
				Border etch = BorderFactory.createEtchedBorder();
				this.setBorder(BorderFactory.createTitledBorder(etch));

				setHorizontalAlignment(SwingConstants.RIGHT);
				setBackground(SystemColor.controlHighlight);
				setOpaque(true);
			}
			return this;
		}
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the start button 's action.
	 */
	class ButtonStartActionListener implements ActionListener{
		public void actionPerformed(ActionEvent e) {
			ErrorCode errorCode = ErrorCode.Success;
			isRunning = true;
			haveData = true;
			for(int i = 0; i < channelCountMax; i++){
				if(channelEnable[i]){
					ChannelStatusStringChange(i, "Running");
				}
			}
			for(int i = 0; i < model.getSize(); i++){
				model.set(i, "");
			}
			
			{		
				errorCode = bfdUdCounterCtrl.Prepare();
				CheckError(errorCode);
				
				try {
					bfdUdCounterCtrl.setEnabled(true);
				} catch (Exception ex) {
					ShowMessage("Sorry, there're some errors occured: " + ex.toString());
					return;
				}
				
				DisplayChannelStatus(displayChannel);
				
				btnConfigure.setEnabled(false);
				btnStart.setEnabled(false);
				btnStop.setEnabled(true);
			}
		}
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the stop button 's action.
	 */
	class ButtonStopActionListener implements ActionListener{
		public void actionPerformed(ActionEvent e) {
			try {
				bfdUdCounterCtrl.setEnabled(false);
			} catch (Exception ex) {
				ShowMessage("Sorry, there're some errors occured: " + ex.toString());
				return;
			}
			isRunning = false;
			btnStart.setEnabled(true);
			btnConfigure.setEnabled(true);
			btnStop.setEnabled(false);
			
			for(int i = 0; i < channelCountMax; i++){
				if(channelEnable[i]){
					ChannelStatusStringChange(i, "Stopped");
				}
			}
		}
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the configure button 's action.
	 */
	class ButtonConfigureActionListener implements ActionListener{
		public void actionPerformed(ActionEvent e) {
			if(bfdUdCounterCtrl.getEnabled()){
				try {
					bfdUdCounterCtrl.setEnabled(false);
				} catch (Exception ex) {
					ShowMessage("Sorry, there're some errors occured: " + ex.toString());
					return;
				}
			}
			haveData = false;
			configureDialog.LoadConfiguration(configure);
			configureDialog.isFirstLoad = false;
			configureDialog.setVisible(true);
		}
	}
	
	class cbxDisplayedChannelItemListener implements ItemListener{

		@Override
		public void itemStateChanged(ItemEvent arg0) {
			// TODO Auto-generated method stub
			if(arg0.getStateChange() == ItemEvent.SELECTED)
            {
				displayChannel = Integer.parseInt(cbxDisplayedChannel.getSelectedItem().toString());
				//System.out.println(" displayChannel==>: " + displayChannel );
            }
			
			if(cbxDisplayedChannel.isEnabled()){
				//System.out.println(" Enabled==>: " + displayChannel );
			    DisplayChannelStatus(displayChannel);
			    DisplayChannelValue(displayChannel);
			}
		}
	}
	
	public class DataReadyEventListener implements BfdCntrEventListener {
		@Override
		public void BfdCntrEvent(Object arg0, BfdCntrEventArgs arg1) {
			// TODO Auto-generated method stub
			int getDataCount = Math.min(arg1.Count, samplesCount);
			eventChannel = arg1.Channel;
			ErrorCode ret = bfdUdCounterCtrl.GetData(arg1.Channel, getDataCount, dataBuf, 0, null, null);
			if(Global.BioFaild(ret)){
				ShowMessage("Sorry, there're some errors occred, ErrorCode: " + ret.toString());
				return;
			}
			CopyChannelData(eventChannel);
			if(eventChannel == displayChannel){
				DisplayChannelValue(displayChannel);
			}
		}
	}
	
	public class StoppedEventListener implements BfdCntrEventListener{
		@Override
		public void BfdCntrEvent(Object arg0, BfdCntrEventArgs arg1) {
			// TODO Auto-generated method stub
			int getDataCount = Math.min(arg1.Count, samplesCount);
			if(getDataCount > 0){
				eventChannel = arg1.Channel;
				ErrorCode ret = bfdUdCounterCtrl.GetData(arg1.Channel, getDataCount, dataBuf, 0, null, null);
				if(Global.BioFaild(ret)){
					ShowMessage("Sorry, there're some errors occred, ErrorCode: " + ret.toString());
					return;
				}
				CopyChannelData(eventChannel);
				if(eventChannel == displayChannel){
					DisplayChannelValue(displayChannel);
				}				
			}
				
			if(bfdUdCounterCtrl.getState() == ControlState.Stopped){
				try {
					bfdUdCounterCtrl.setEnabled(false);
				} catch (Exception ex) {
					ShowMessage("Sorry, there're some errors occured: " + ex.toString());
					return;
				}
				
				isRunning = false;
				btnStart.setEnabled(true);
				btnConfigure.setEnabled(true);
				btnStop.setEnabled(false);
			}
			
			ChannelStatusStringChange(eventChannel, "Stopped");		
			DisplayChannelStatus(displayChannel);
		}	
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the main frame's closing event.
	 */
	class WindowCloseActionListener extends WindowAdapter{
		@Override
		public void windowClosing(WindowEvent e) {
			if (bfdUdCounterCtrl.getEnabled()) {
				try {
					bfdUdCounterCtrl.setEnabled(false);
					bfdUdCounterCtrl.Cleanup();
				} catch (Exception ex) {
					// do nothing here.
					// ShowMessage("Sorry, there're some errors occurred: " + ex.toString());
				}
			}
		}
	}
}

/**
 * 
 * @author Administrator 
 * Class Description: this class is use to transfer parameter to the main frame !
 */
class ConfigureParameter {
	public String deviceName;
	public int sectionLength;
	public int sectionCount;
	public SignalDrop sampleClkSrc;
	public CountingType cntType;
	
	public SignalDrop triggerSource;
	public TriggerAction triggerAction;	
	public ActiveSignal triggerEdge;
	public double triggerLevel;
	
	public boolean[] channelEnabled;
	
	public int configureMode; //load profile --> 1  /  manual configure --> 0
}

File: SDK Examples/Examples/Qt/Counter_BufferedUpDownCounter/bufferedupdowncounter.cpp
#include <QPalette>
#include <QMessageBox>
#include <QListWidget>
#include <QListWidgetItem>
#include "bufferedupdowncounter.h"

BufferedUpdownCounter::BufferedUpdownCounter(QWidget *parent, Qt::WindowFlags flags)
	: QDialog(parent, flags)
{
	ui.setupUi(this);
	//Set the minimum and close button of the main frame.
	this->setWindowFlags(Qt::WindowFlags(Qt::WindowSystemMenuHint | Qt::WindowTitleHint | Qt::WindowCloseButtonHint));

	//Set the background image for the main frame.
	ui.bkgrndImage->setAutoFillBackground(true);
    QPixmap pixMap(":/Counter_BufferedUpDownCounter/Resources/Background.png");
	QPalette backPalette;
	backPalette.setBrush(ui.bkgrndImage->backgroundRole(),QBrush(pixMap));
	ui.bkgrndImage->setPalette(backPalette);

    dataBuf = NULL;
    channelEnable = NULL;
    chs_dataBuf = NULL;
    chs_Status = NULL;

    bfdUdCounterCtrl = BufferedUdCounterCtrl::Create();
    bfdUdCounterCtrl->addDataReadyHandler(OnBfdUdCounterEvent, this);
    bfdUdCounterCtrl->addStoppedHandler(OnBfdUdCounterStoppedEvent, this);

	connect(ui.btnStart, SIGNAL(clicked()), this, SLOT(ButtonStartClicked()));
	connect(ui.btnStop, SIGNAL(clicked()), this, SLOT(ButtonStopClicked()));
	connect(ui.btnConfig, SIGNAL(clicked()), this, SLOT(ButtonConfigureClicked()));	
    connect(this, SIGNAL(refreshUI(int)), this, SLOT(onRefreshUI(int)));
    connect(this, SIGNAL(refreshStatus(int)), this, SLOT(onRefreshStatus(int)));
    connect(ui.cbxDisplayedChannel, SIGNAL(currentIndexChanged(int)), this, SLOT(onDisplayedChannelChanged()));
}

BufferedUpdownCounter::~BufferedUpdownCounter()
{
    if(dataBuf != NULL){
        delete[]dataBuf;
        dataBuf = NULL;
    }
    if (channelEnable != NULL ){
        delete []channelEnable;
        channelEnable = NULL;
    }
    if(chs_Status != NULL){
        delete []chs_Status;
        chs_Status = NULL;
    }
    if(chs_dataBuf != NULL){
        for(int i = 0; i < channelCountMax; i++){
            if(chs_dataBuf[i] != NULL)
            delete []chs_dataBuf[i];
        }
        delete chs_dataBuf;
    }
    if(bfdUdCounterCtrl != NULL){
        bfdUdCounterCtrl->Dispose();
        bfdUdCounterCtrl= NULL;
    }
}

void BufferedUpdownCounter::Initialize() {
	//set the title of the form.
    this->setWindowTitle(tr("Buffered Up Down Counter - Run(") + configure.deviceName + tr(")"));
    ui.txtChannelState->clear();
    ConfigureDevice();
	
	ui.btnStart->setEnabled(true);
	ui.btnStop->setEnabled(false);
	ui.btnConfig->setEnabled(true);

    ui.cntrValueList->setEnabled(true);
    ui.txtChannelState->setEnabled(true);

    //initialize ECisRunning
    isRunning = false;
    haveData = false;

    ui.cntrValueList->clear();
}

void BufferedUpdownCounter::ConfigureDevice() {
    ErrorCode errorCode = Success;
    // Setting Device
    std::wstring description = configure.deviceName.toStdWString();
    DeviceInformation selected(description.c_str());
    errorCode = bfdUdCounterCtrl->setSelectedDevice(selected);
    CheckError(errorCode);

    //Get Channel Count Max
    channelCountMax = bfdUdCounterCtrl->getFeatures()->getChannelCountMax();
    channelEnable = new int8[channelCountMax];
    memset(channelEnable, 0, sizeof(int8)* channelCountMax);

    chs_Status = new QString[channelCountMax];

    //configure parameters mode (Profile configure/Manual configure)
    if(!configure.configMode){
        errorCode = bfdUdCounterCtrl->LoadProfile(configure.profilePath);
        CheckError(errorCode);
        bfdUdCounterCtrl->getChannelMap(channelCountMax, channelEnable);
        sectionLength = bfdUdCounterCtrl->getRecord()->getSectionLength();
        sectionCount = bfdUdCounterCtrl->getRecord()->getSectionCount();
    }else{
        memcpy(channelEnable, configure.channelEnable, sizeof(int8) * channelCountMax);
        sectionLength = configure.sectionLength;
        sectionCount = configure.sectionCount;

        errorCode = bfdUdCounterCtrl->setChannelMap(channelCountMax,configure.channelEnable);
        CheckError(errorCode);

        errorCode = bfdUdCounterCtrl->setSampleClkSrc(configure.clkSource);
        CheckError(errorCode);

        Record* record = bfdUdCounterCtrl->getRecord();
        errorCode = record->setSectionLength(configure.sectionLength);
        CheckError(errorCode);
        errorCode = record->setSectionCount(configure.sectionCount);
        CheckError(errorCode);

        for(int i = 0; i < channelCountMax; i++){
            if(configure.channelEnable[i] & 0x1){
                errorCode = bfdUdCounterCtrl->getChannels()->getItem(i).setCountingType(configure.cntType);
                CheckError(errorCode);
            }
        }

        Trigger* trigger = bfdUdCounterCtrl->getTrigger();
        if(trigger){
            errorCode = trigger->setSource(configure.trgSource);
            CheckError(errorCode);
            errorCode = trigger->setAction(configure.trgAction);
            CheckError(errorCode);
            errorCode = trigger->setEdge(configure.trgEdge);
            CheckError(errorCode);
            errorCode = trigger->setLevel(configure.trgLevel);
            CheckError(errorCode);
        }
    }

    samples = sectionLength;
    if(sectionCount){
        samples *= configure.sectionCount;
    }
    dataBuf = new int32[samples];
    memset(dataBuf, 0, sizeof(int32) * samples);

    chs_dataBuf = new int32*[channelCountMax];
    for(int i = 0; i < channelCountMax; i++ ){
        chs_dataBuf[i] = new int32[samples];
        memset(chs_dataBuf[i], 0, sizeof(int32) * samples);
    }

    if ((dataBuf == NULL) || (chs_dataBuf == NULL) )
    {
        QMessageBox::information(this, tr("Warning Information"),
            tr("Sorry! Error in allocating memory...."));
        return;
    }

    ui.cbxDisplayedChannel->clear();
    //enum enabled channels
    for(int i = 0; i < channelCountMax; i++){
        if(channelEnable[i] & 0x1){
            ui.cbxDisplayedChannel->addItem(QString("%1").arg(i));
        }
    }
    ui.cbxDisplayedChannel->setCurrentIndex(0);

    // Setting Status is Ready.
    for(int i = 0; i < channelCountMax; i++){
       if(channelEnable[i] == 1){
           ChannelStatusStringChange(i, "Ready");
       }
    }
    displayedChannel = ui.cbxDisplayedChannel->currentText().toInt();
    DisplayChannelStatus(displayedChannel);
}

void BufferedUpdownCounter::ButtonStartClicked() {
	ErrorCode errorCode = Success;
    ui.cntrValueList->clear();

    // Setting Status is Running.
    for(int i = 0; i < channelCountMax; i++){
       if(channelEnable[i] == 1){
           ChannelStatusStringChange(i, "Running");
       }
    }
    DisplayChannelStatus(displayedChannel);

    errorCode = bfdUdCounterCtrl->Prepare();
    CheckError(errorCode);
    errorCode = bfdUdCounterCtrl->setEnabled(true);
    CheckError(errorCode);
    isRunning = true;
    haveData = true;

    ui.btnStop->setEnabled(true);
    ui.btnStart->setEnabled(false);

	ui.btnConfig->setEnabled(false);
}

void BufferedUpdownCounter::ButtonStopClicked() {
	//Stop event counting
	ErrorCode errorCode = Success;
    errorCode = bfdUdCounterCtrl->setEnabled(false);
	CheckError(errorCode);
    isRunning = false;

	ui.btnStart->setEnabled(true);
	ui.btnStop->setEnabled(false);
	ui.btnConfig->setEnabled(true);

    for(int i = 0; i < channelCountMax; i++){
       if(channelEnable[i] == 1){
           ChannelStatusStringChange(i, "Stopped");
       }
    }
    DisplayChannelStatus(displayedChannel);
}

void BufferedUpdownCounter::ButtonConfigureClicked() {
    configDialog->RefreshConfigureParameter();
    int dialogResult = configDialog->exec();
	if (dialogResult == QDialog::Accepted)
	{
		configure = configDialog->GetConfigureParameter();
		Initialize();
    }
    haveData = false;
}

void BufferedUpdownCounter::CheckError(ErrorCode errorCode) {
    if (BioFailed(errorCode))
	{
		QString message = tr("Sorry, there are some errors occurred, Error Code: 0x") +
			QString::number(errorCode, 16).right(8).toUpper();
		QMessageBox::information(this, "Warning Information", message);
	}
}

void BufferedUpdownCounter::OnBfdUdCounterEvent(void *sender, BfdCntrEventArgs *args, void *userParam)
{
    ErrorCode ret = Success;
    int32 returnedCount = 0;
    BufferedUpdownCounter* uParam = (BufferedUpdownCounter*)userParam;
    BufferedEventCounterCtrl* ctrl = (BufferedEventCounterCtrl*)sender;
    int32 getDataCount = qMin(args->Count, uParam->samples);
    uParam->eventChannel = args->Channel;
    if(getDataCount){
        ret = ctrl->GetData(args->Channel, getDataCount, uParam->dataBuf, 0, &returnedCount, NULL);
        emit uParam->refreshUI(args->Channel);
    }
}

void BufferedUpdownCounter::onRefreshUI(int channel)
{ 
    CopyChannelData(channel);
    if(eventChannel == displayedChannel){
        //display channel value.
        DisplayChannelValue(displayedChannel);
    }
}

void BufferedUpdownCounter::OnBfdUdCounterStoppedEvent(void *sender, BfdCntrEventArgs *args, void *userParam)
{
    ErrorCode ret = Success;
    int32 returnedCount = 0;
    BufferedUpdownCounter* uParam = (BufferedUpdownCounter*)userParam;
    BufferedEventCounterCtrl* ctrl = (BufferedEventCounterCtrl*)sender;
    int32 getDataCount = qMin(args->Count, uParam->samples);
    uParam->eventChannel = args->Channel;
    if(getDataCount > 0){
        ret = ctrl->GetData(args->Channel, getDataCount, uParam->dataBuf, 0, &returnedCount, NULL);
        emit uParam->refreshUI(args->Channel);
    }

    emit uParam->refreshStatus(args->Channel);
}

void BufferedUpdownCounter::onRefreshStatus(int channel)
{
    if(bfdUdCounterCtrl->getState() == Stopped){
        ErrorCode errorCode = bfdUdCounterCtrl->setEnabled(FALSE);
        CheckError(errorCode);
        ui.btnStart->setEnabled(true);
        ui.btnStop->setEnabled(false);
        ui.btnConfig->setEnabled(true);
        isRunning = false;
       }
    ChannelStatusStringChange(channel, "Stopped");
    DisplayChannelStatus(displayedChannel);
}

void BufferedUpdownCounter::onDisplayedChannelChanged()
{
    displayedChannel = ui.cbxDisplayedChannel->currentText().toInt();

    DisplayChannelStatus(displayedChannel);
    DisplayChannelValue(displayedChannel);
}

void BufferedUpdownCounter::ChannelStatusStringChange(int32 channel, QString string)
{
    chs_Status[channel] = string;
}

void BufferedUpdownCounter::DisplayChannelStatus(int channel)
{
    int ch = channel;
    ui.txtChannelState->setText(chs_Status[ch]);
}

void BufferedUpdownCounter::DisplayChannelValue(int channel)
{
    int ch = channel;
    ui.cntrValueList->clear();

    int listCount = 0;
    if(sectionLength >= 10){
        listCount = 10;
    }else{
        listCount = sectionLength;
    }

    QListWidgetItem* item = NULL;
    if(haveData || isRunning){
        for(int i = 0; i < listCount; i++){
            item = new QListWidgetItem(QString("%1").arg(chs_dataBuf[ch][i]), ui.cntrValueList);
            item->setTextAlignment(Qt::AlignLeft);
        }
    }
}

void BufferedUpdownCounter::CopyChannelData(int channel)
{
    int ch = channel;
    memcpy(chs_dataBuf[ch], dataBuf, sizeof(int32)*samples);
}
